### DISCUSSION TITLE
webpack cant insert code to RuntimeModule, like mainTemplate hooks in webpack4

### DISCUSSION THREAD

**Discussion by lanmch** [webpack/webpack#15011] [Q&A]
# Bug report

**What is the current behavior?**

As the webpack5 docs, some mainTemplate hooks has been removed. But in my project with webpack4, we use the mainTemplate.hooks.bootstrap and mainTemplate.hooks.requireEnsure to inject some js to the final file. But now the mainTemplate feature is not exist, and the webpack5 docs now show how to change it. And from the webpack5 source code, I cant find any new hooks.

my plugin code
```javascript
apply(compiler) {
    if (compiler.hooks) {
      compiler.hooks.compilation.tap('RetryPluguin', compilation => {
        compilation.mainTemplate.hooks.bootstrap.tap(
          'RetryPluginAddCustomLogic',
          (source, chunk, hash, moduleIdExpression) => {
            const flagName = JSON.stringify(`__manifest_executed_${chunk.ids.join('_')}_${hash}__`);
            const newSource = makeTagRetry(source, ensureExecOnceToBeInject, getExecOnceExecScript(flagName));
            return newSource;
          },
        );
        compilation.mainTemplate.hooks.requireEnsure.tap(
          'RetryPluginAddCustomLogic',
          (source, chunk, hash, moduleIdExpression) => {
            let newSource = source;
            console.log('source: ', source); // nothing
            newSource = makeTagRetry(newSource, scriptToBeInject, scriptRetry);
            newSource = makeTagRetry(newSource, linkToBeInject, cssRetry);
            return newSource;
          },
        );
      });
    }
  }
```
**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

**Other relevant information:**
webpack version:  5.51.1
Node.js version: 14.16
Operating System: macos
Additional tools:

**Comment by alexander-akait** [comment]
Please use discussion for questions, there are no bugs, you already have https://github.com/webpack/webpack/discussions/14995, no need duplicate

**Comment by lanmch** [comment]
您好，我已收到您的邮件，我会尽快给您回复。

**Reply by alexander-akait** [reply]
Please use English
